File: kotlin-project/factory.kt
interface Animal {
    fun speak(): String
}

class Dog : Animal {
    override fun speak(): String {
        return "Woof!"
    }
}

class Cat : Animal {
    override fun speak(): String {
        return "Meow!"
    }
}

class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

fun main() {
    val factory = AnimalFactory()

    val dog = factory.createAnimal("dog")
    if (dog != null) {
        println("Dog says: ${dog.speak()}")
    }

    val cat = factory.createAnimal("cat")
    if (cat != null) {
        println("Cat says: ${cat.speak()}")
    }

    val unknown = factory.createAnimal("bird")
    if (unknown == null) {
        println("Unknown animal type!")
    }
}

File: cpp-project/factory.cpp
#include <iostream>
#include <memory>
#include <string>

class Animal {
public:
    virtual std::string speak() const = 0; // 순수 가상 함수
    virtual ~Animal() = default; // 가상 소멸자
};

class Dog : public Animal {
public:
    std::string speak() const override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    std::string speak() const override {
        return "Meow!";
    }
};

class AnimalFactory {
public:
    std::unique_ptr<Animal> createAnimal(const std::string& animalType) {
        if (animalType == "dog") {
            return std::make_unique<Dog>();
        } else if (animalType == "cat") {
            return std::make_unique<Cat>();
        } else {
            return nullptr;
        }
    }
};

int main() {
    AnimalFactory factory;

    auto dog = factory.createAnimal("dog");
    if (dog) {
        std::cout << "Dog says: " << dog->speak() << std::endl;
    }

    auto cat = factory.createAnimal("cat");
    if (cat) {
        std::cout << "Cat says: " << cat->speak() << std::endl;
    }

    auto unknown = factory.createAnimal("bird");
    if (!unknown) {
        std::cout << "Unknown animal type!" << std::endl;
    }

    return 0;
}